### PR TITLE
Extend the module interface ops

### PR DESCRIPTION
**Pull request by ranj063**
Extend the module interface to add the bind/unbind ops and also make it easier to convert the copier to use the module adapter

### CODE REVIEW

File: src/audio/module_adapter/module/generic.c [hunk 1]
```diff
@@ -436,3 +436,21 @@ int module_set_configuration(struct processing_module *mod,
...
 	return ret;
 }
+
+int module_bind(struct processing_module *mod, void *data)
+{
+	struct module_data *md = &mod->priv;
+
+	if (md->ops->bind)
+		return md->ops->bind(mod, data);
+	return 0;
```
Comment (kv2019i) [inline src/audio/module_adapter/module/generic.c:446]: L445-446 extra line?

File: src/include/sof/audio/module_adapter/module/generic.h [hunk 2]
```diff
@@ -235,6 +235,8 @@ int module_set_configuration(struct processing_module *mod,
 			     enum module_cfg_fragment_position pos, size_t data_offset_size,
 			     const uint8_t *fragment, size_t fragment_size, uint8_t *response,
 			     size_t response_size);
+int module_bind(struct processing_module *mod, void *data);
```
Comment (kv2019i) [inline src/include/sof/audio/module_adapter/module/generic.h:238]: Could we have some more specific type than "void *data"? Given this is already "generic.h", we should know what is passed. If there's some other variant of data passed, then that should be probably another ops.

File: src/include/sof/audio/module_adapter/module/module_interface.h [hunk 3]
```diff
@@ -143,6 +143,14 @@ struct module_interface {
 	 * free in .free(). This should free all memory allocated during module initialization.
 	 */
 	int (*free)(struct processing_module *mod);
+	/**
+	 * Module specific bind procedure, called when modules are bound with each other
```
Comment (kv2019i) [inline src/include/sof/audio/module_adapter/module/module_interface.h:213]: Should have some documentation on what "data" is and what the conditions are (is the new state reflected in module internals or not).

**Review by kv2019i — COMMENTED** [review]
Some comments inline.

File: src/include/sof/audio/module_adapter/module/module_interface.h [hunk 4]
```diff
@@ -65,6 +65,72 @@ struct output_stream_buffer {
...
 };
 
+/**
+ * \struct module_endpoint_ops
+ * \brief Ops relevant only for the endpoint devices such as the host copier or DAI copier.
+ *	  Other modules should not implement these.
+ */
+struct module_endpoint_ops {
+	/**
+	 * Returns total data processed in number bytes.
```
[the remark this answers is not in the document]
Comment (ranj063) [reply src/include/sof/audio/module_adapter/module/module_interface.h:75]: @lyakh if you really want these changed, i'd like to do them as a pre-requisite. I'm only moving them from the comp driver ops

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

File: src/include/sof/audio/module_adapter/module/module_interface.h [hunk 5]
```diff
@@ -65,6 +65,72 @@ struct output_stream_buffer {
...
+	 * @param dev Component device
+	 * @param stream_no Index of input/output stream
+	 * @param input Selects between input (true) or output (false) stream direction
+	 * @return total data processed if succeeded, 0 otherwise.
+	 */
+	uint64_t (*get_total_data_processed)(struct comp_dev *dev, uint32_t stream_no, bool input);
```
[the remark this answers is not in the document]
Comment (btian1) [reply src/include/sof/audio/module_adapter/module/module_interface.h:81]: I agree on this, however, current code implementation is uint64_t for modules, let's keep this right now, and later can use one PR to change to uint32_t.

File: src/include/sof/audio/module_adapter/module/module_interface.h [hunk 6]
```diff
@@ -65,6 +65,72 @@ struct output_stream_buffer {
...
+	/**
+	 * Retrieves component rendering position.
+	 * @param dev Component device.
+	 * @param posn Receives reported position.
+	 */
+	int (*position)(struct comp_dev *dev, struct sof_ipc_stream_posn *posn);
```
[the remark this answers is not in the document]
Comment (btian1) [reply src/include/sof/audio/module_adapter/module/module_interface.h:87]: aligned with copier_position.

**Review by btian1 — APPROVED** [review]

**Pull request closed without merging**